sdks/python/apache_beam/runners/dataflow/dataflow_runner.py

```diff
@@ -118,13 +118,15 @@ class DataflowRunner(PipelineRunner):
   from apache_beam.runners.dataflow.ptransform_overrides import CreatePTransformOverride
   from apache_beam.runners.dataflow.ptransform_overrides import JrhReadPTransformOverride
   from apache_beam.runners.dataflow.ptransform_overrides import ReadPTransformOverride
+  from apache_beam.runners.dataflow.ptransform_overrides import ReadBigQuerySourcePTransformOverride
   from apache_beam.runners.dataflow.ptransform_overrides import NativeReadPTransformOverride
 
   # These overrides should be applied before the proto representation of the
   # graph is created.
   _PTRANSFORM_OVERRIDES = [
       CombineValuesPTransformOverride(),
       NativeReadPTransformOverride(),
+      ReadBigQuerySourcePTransformOverride(),
```

Comment:
The problem with putting this as an override here is that it would be called *after* the beam protos are created. 

Instead of implementing this as a (Dataflow-only) override, it'd be better to change `beam.io.bigquery.BigQuerySource` into a function that returns a `ReadFromBigquery` transform. 

We could either remove the native `BigQuerySource` altogether, or add a (temporary) keyword argument `use_native_source` to our new `BigQuerySource` function to branch on as an escape hatch.

Reply:
alright, I'll do that instead then.

Reply:
.. an issue with this, though, is that BigQuerySource is wrapped by a Read transform. So we'd have `Read(BigQuerySource(...))` - and so we would have then a `Read(ReadFromBigQuery(...))`, which would not work. We can also return the `_CustomBigQuerySource`, but it would be missing the clean up after the export jobs. Thoughts?

Reply:
Furthermore, BQSource is only useful on Dataflow (it does have a local workaround implementation for DirectRunner, which we can also override if it makes sense..)

Reply:
Seems like we are creating the proto pipeline after applying the overrides: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/dataflow/dataflow_runner.py#L482

So overrides will be reflected in the proto ?

Reply:
You're right, the proto is created later. I still think it would be cleaner to construct the right pipeline from the start than mutate it later with overrides. Let's patch up Read to accept PTransforms just as we do for Write https://github.com/apache/beam/blob/release-2.18.0/sdks/python/apache_beam/io/iobase.py#L951 (I thought this was already done.)

Reply:
Working on https://github.com/apache/beam/pull/12782 to do the correct override